Move implicit mapping out of CSSPrimitiveValue

darinadler opened the pull request:
#### d182c81af891b767565c56e55eb3d7811a5ed745
<pre>
Move implicit mapping out of CSSPrimitiveValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=251567">https://bugs.webkit.org/show_bug.cgi?id=251567</a>
rdar://104946191

Reviewed by NOBODY (OOPS!).

Our automatically generated style building code depends on implicitly
converting CSSValue objects to various destination types. That was done
by a conversion operator in CSSPrimitiveValue, but it&apos;s better if the
conversion operator is separate, so the conversion can&apos;t be done by
accident in normal uses of CSSPrimitiveValue. At the very least, such
unintentional conversions can lead to confusing compiler errors.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::setDisplay): Use fromCSSValue.

* Source/WebCore/css/CSSPrimitiveValue.h: Deleted all the conversion
operators, including the ones for numeric types, and the template.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::fromCSSValue): Added.
(WebCore::TypeDeducingCSSValueMapper::TypeDeducingCSSValueMapper): Added.
(WebCore::TypeDeducingCSSValueMapper::operator TargetType const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator const CSSPrimitiveValue&amp; const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator int const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator float const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator double const): Added.
(WebCore::TypeDeducingCSSValueMapper::numericValue const): Added.
(WebCore::fromCSSValueDeducingType): Added.
(WebCore::fromCSSValue): Replaced CSSPrimitiveValue specialization for LineClampValue with
fromCSSValue specialization.
(WebCore::CSSPrimitiveValue::operator ColumnSpan const): Deleted. This was not needed because we
no longer allow a number for column-span; when we removed support in the parser we missed this.
(WebCore::fromCSSValueID): Remove unneeded specializations for OptionSet&lt;HangingPunctuation&gt;,
OptionSet&lt;TextDecorationLine&gt;, and OptionSet&lt;TouchAction&gt;, using
DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS instead.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillClip): Use fromCSSValue.
(WebCore::CSSToStyleMap::mapFillComposite): Ditto.
(WebCore::CSSToStyleMap::mapFillBlendMode): Ditto.
(WebCore::CSSToStyleMap::mapFillOrigin): Ditto.
(WebCore::CSSToStyleMap::mapFillXPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillYPosition): Ditto.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Updated since the underlying
type for BoxOrient is bool instead of uint8_t.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext.generate_heading): Tweak comment.
(GenerateCSSPropertyNames): Use fromCSSValueDeducingType.

* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::setTextAlignmentForChangedBaseWritingDirection): Use propertyAsValueID,
CSSValueID, and nameString.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const): Use propertyAsValueID
and fromCSSValueID.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;): Removed support for HangingPunctuation::None.

* Source/WebCore/rendering/style/RenderStyleConstants.h: Converted all two-value enumerations
to be based on bool instead of uint8_t. Removed HangingPunctuation::None.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationLine): Use fromCSSValue.
(WebCore::Style::BuilderConverter::convertTextAlign): Ditto.
(WebCore::Style::BuilderConverter::convertTextAlignLast): Ditto.
(WebCore::Style::BuilderConverter::convertPathOperation): Ditto.
(WebCore::Style::BuilderConverter::convertResize): Ditto.
(WebCore::Style::BuilderConverter::convertTextUnderlinePosition): Ditto.
(WebCore::Style::BuilderConverter::convertReflection): Ditto.
(WebCore::Style::BuilderConverter::convertShapeValue): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapType): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapAlign): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapStop): Ditto.
(WebCore::Style::BuilderConverter::convertTouchAction): Ditto.
(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData): Ditto.
(WebCore::Style::BuilderConverter::convertSpeakAs): Ditto.
(WebCore::Style::BuilderConverter::convertHangingPunctuation): Ditto.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueDirection): Ditto.
(WebCore::Style::BuilderCustom::applyValueVerticalAlign): Ditto.
(WebCore::Style::BuilderCustom::applyValueListStyleType): Ditto.
(WebCore::Style::BuilderCustom::applyValueOutlineStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueWritingMode): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextOrientation): Ditto.
(WebCore::Style::BuilderCustom::applyValueDisplay): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextEmphasisStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueCursor): Ditto.

* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::colorInterpolation const): Ditto.
</pre>

nt1m commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -107,12 +186,11 @@ DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS
-template<> inline CSSPrimitiveValue::operator ColumnSpan() const
+template<> inline ImplicitValueMapper::operator ColumnSpan() const
 {
-    // Map 1 to none for compatibility reasons.
```
Can we keep this comment?

darinadler replied:
I would be wiling to keep it, but hear my out about why I don’t think it’s a good comment:

1) The comment doesn’t say enough. Compatibility with what? There’s no reference here to code, a specification, a particular website, or even a test.
2) The comment is in the wrong place. Mapping the 1 value when converting the enumeration would be only part of what’s needed, because we also need to at least parse that value. The whole point of these mapping functions is that they only work when the value is already known to be one of the correct ones, and the code that guarantees that is the CSS parser, and to a lesser extent the CSS computed style extractor.
3) Comments like this don’t really help much without tests. And see below to see what that has led to!

I think the real issue is that this is dead code. I don’t think we do map 1 to none for compatibility reasons! We certainly don’t do it in the parsing code for `column-span` nor in the computed style extraction. I’ll leave this poor comment in for now at your request, and then I’ll return and prove that this is not needed and delete it.

nt1m replied:
Yes, removing the code + the comment entirely in a separate commit seems better.

Removing just the comment leaves the risk that this code is never actually removed if it is deprecated.

nt1m commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -107,12 +186,11 @@ DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS
+    if (downcast<CSSPrimitiveValue>(m_value).doubleValue() == 1)
         return ColumnSpan::None;
-    return fromCSSValueID<ColumnSpan>(valueID());
+    return fromCSSValueID<ColumnSpan>(downcast<CSSPrimitiveValue>(m_value).valueID());
```
Can we store the result of the downcast in a variable?

darinadler replied:
I’d prefer not to, but I suppose it really doesn’t matter since this code will be deleted soon enough.

nt1m commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -50,6 +50,83 @@
+    return value;
+}
+
+inline ImplicitValueIDMapper implicitMapper(CSSValueID value)
```
Not a huge fan of the name `implicitMapper`. I think it should be named what it actually maps to, or `mapTo[...]`.

darinadler replied:
I understand that you don’t like the name, and I am not all that fond of it myself, and open to changing it. But I can’t change it to name what it maps to, because the entire point of this is that it maps to whatever type you pass it to. That’s the feature that I am taking out of `CSSPrimitiveValue` and moving into this mapper class. So it‘s more like "map automatically" or "map based on what type you assign this to". I would absolutely be willing to rename this something else, but the suggestions you made would be for something that maps to a specific type, not automatically to whatever it’s assigned to.

darinadler replied:
I’ll land and we can globally replace to a better name: `mapAutomatically`, `mapImplicitly`, `convertAutomatically`, `convertImplicitly`. I also think we should use this as little as possible. It’s important mainly for automatically generated code, but we could also just remove this if someone can convert all that code to not need it.

darinadler replied:
We have `fromCSSValueID<>`, so maybe we could add `fromCSSValue<>` and this could be named `fromCSSValueDeducingType`?

nt1m replied:
Yes, those names all seem better than implicitMapper.

nt1m commented on Source/WebCore/css/process-css-properties.py:
```diff
@@ -2575,17 +2575,21 @@ def _generate_lookup_functions(self, *, to):
                     *nextCharacter++ = character;
                 }
+                if (!sawHyphen)
+                    return nameString(id);
```
These changes look unrelated

darinadler replied:
Oh yes, they are. It’s performance optimization that I included intentionally, forgot to mention in the change log, and am happy to land separately.

nt1m replied:
seems fine

darinadler committed:
Move implicit mapping out of CSSPrimitiveValue
https://bugs.webkit.org/show_bug.cgi?id=251567
rdar://104946191

Reviewed by Tim Nguyen.

Our automatically generated style building code depends on implicitly
converting CSSValue objects to various destination types. That was done
by a conversion operator in CSSPrimitiveValue, but it's better if the
conversion operator is separate, so the conversion can't be done by
accident in normal uses of CSSPrimitiveValue. At the very least, such
unintentional conversions can lead to confusing compiler errors.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::setDisplay): Use fromCSSValue.

* Source/WebCore/css/CSSPrimitiveValue.h: Deleted all the conversion
operators, including the ones for numeric types, and the template.

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::fromCSSValue): Added.
(WebCore::TypeDeducingCSSValueMapper::TypeDeducingCSSValueMapper): Added.
(WebCore::TypeDeducingCSSValueMapper::operator TargetType const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator const CSSPrimitiveValue& const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned short const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator int const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator unsigned const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator float const): Added.
(WebCore::TypeDeducingCSSValueMapper::operator double const): Added.
(WebCore::TypeDeducingCSSValueMapper::numericValue const): Added.
(WebCore::fromCSSValueDeducingType): Added.
(WebCore::fromCSSValue): Replaced CSSPrimitiveValue specialization for LineClampValue with
fromCSSValue specialization.
(WebCore::CSSPrimitiveValue::operator ColumnSpan const): Deleted. This was not needed because we
no longer allow a number for column-span; when we removed support in the parser we missed this.
(WebCore::fromCSSValueID): Remove unneeded specializations for OptionSet<HangingPunctuation>,
OptionSet<TextDecorationLine>, and OptionSet<TouchAction>, using
DEFINE_TO_FROM_CSS_VALUE_ID_FUNCTIONS instead.

* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillClip): Use fromCSSValue.
(WebCore::CSSToStyleMap::mapFillComposite): Ditto.
(WebCore::CSSToStyleMap::mapFillBlendMode): Ditto.
(WebCore::CSSToStyleMap::mapFillOrigin): Ditto.
(WebCore::CSSToStyleMap::mapFillXPosition): Ditto.
(WebCore::CSSToStyleMap::mapFillYPosition): Ditto.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Updated since the underlying
type for BoxOrient is bool instead of uint8_t.

* Source/WebCore/css/process-css-properties.py:
(GenerationContext.generate_heading): Tweak comment.
(GenerateCSSPropertyNames): Use fromCSSValueDeducingType.

* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::setTextAlignmentForChangedBaseWritingDirection): Use propertyAsValueID,
CSSValueID, and nameString.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const): Use propertyAsValueID
and fromCSSValueID.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<): Removed support for HangingPunctuation::None.

* Source/WebCore/rendering/style/RenderStyleConstants.h: Converted all two-value enumerations
to be based on bool instead of uint8_t. Removed HangingPunctuation::None.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationLine): Use fromCSSValue.
(WebCore::Style::BuilderConverter::convertTextAlign): Ditto.
(WebCore::Style::BuilderConverter::convertTextAlignLast): Ditto.
(WebCore::Style::BuilderConverter::convertPathOperation): Ditto.
(WebCore::Style::BuilderConverter::convertResize): Ditto.
(WebCore::Style::BuilderConverter::convertTextUnderlinePosition): Ditto.
(WebCore::Style::BuilderConverter::convertReflection): Ditto.
(WebCore::Style::BuilderConverter::convertShapeValue): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapType): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapAlign): Ditto.
(WebCore::Style::BuilderConverter::convertScrollSnapStop): Ditto.
(WebCore::Style::BuilderConverter::convertTouchAction): Ditto.
(WebCore::Style::BuilderConverter::convertSelfOrDefaultAlignmentData): Ditto.
(WebCore::Style::BuilderConverter::convertSpeakAs): Ditto.
(WebCore::Style::BuilderConverter::convertHangingPunctuation): Ditto.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueDirection): Ditto.
(WebCore::Style::BuilderCustom::applyValueVerticalAlign): Ditto.
(WebCore::Style::BuilderCustom::applyValueListStyleType): Ditto.
(WebCore::Style::BuilderCustom::applyValueOutlineStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueWritingMode): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextOrientation): Ditto.
(WebCore::Style::BuilderCustom::applyValueDisplay): Ditto.
(WebCore::Style::BuilderCustom::applyValueTextEmphasisStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueCursor): Ditto.

* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::colorInterpolation const): Ditto.

Canonical link: https://commits.webkit.org/259856@main
